zproject/computed_settings.py

```diff
@@ -1181,6 +1183,25 @@ def file_handler(
         path = f"/etc/zulip/saml/idps/{idp_name}.crt"
     idp_dict["x509cert"] = get_from_file_if_exists(path)
 
+
+def ensure_dict_path(d: dict[str, Any], keys: list[str]) -> None:
+    for key in keys:
+        if key not in d:
+            d[key] = {}
+        d = d[key]
+
+
+# Merge SOCIAL_AUTH_SYNC_CUSTOM_ATTRS_DICT into SOCIAL_AUTH_SYNC_ATTRS_DICT.
+# This is compat code for the original SOCIAL_AUTH_CUSTOM_ATTRS_DICT setting.
+# TODO: We can delete this if we want to just remove the old setting.
+for subdomain, dict_for_subdomain in SOCIAL_AUTH_SYNC_CUSTOM_ATTRS_DICT.items():
+    for backend_name, custom_attrs_map in dict_for_subdomain.items():
+        ensure_dict_path(SOCIAL_AUTH_SYNC_ATTRS_DICT, [subdomain, backend_name])
+        for custom_attr_name, source_attr_name in custom_attrs_map.items():
+            SOCIAL_AUTH_SYNC_ATTRS_DICT[subdomain][backend_name][f"custom__{custom_attr_name}"] = (
+                source_attr_name
+            )
```

Comment:
I think it's helpful to have the compatible option and remove it later. You can do a `TODO/compatibility` comment for future removal, and add an "Upgrade notes" entry documenting that. Perhaps:

> - The  `SOCIAL_AUTH_SYNC_CUSTOM_ATTRS_DICT` setting is deprecated in favor of the more general `SOCIAL_AUTH_SYNC_ATTRS_DICT` setting structure, but still works in this release for a smooth upgrade experience. The new setting supports synchronizing `role`, and otherwise functions like the old one, except Zulip custom profile fields are referred to with the prefix `custom__`. See the updated comment documentation for details.